src/govuk/components/modal-dialogue/modal-dialogue.js

```diff
@@ -0,0 +1,186 @@
+  this.$dialogBox = $module.querySelector('dialog')
+  this.$container = document.documentElement
+
+  // Check for browser support
+  this.hasNativeDialog = 'showModal' in this.$dialogBox
+
+  // Allowed focussable elements
+  this.focussable = [
+    'button',
+    '[href]',
+    'input',
+    'select',
+    'textarea'
+  ]
+}
+
+// Initialize component
+ModalDialogue.prototype.init = function (options) {
+  this.options = options || {}
+
+  this.open = this.handleOpen.bind(this)
+  this.close = this.handleClose.bind(this)
+  this.focus = this.handleFocus.bind(this)
+  this.focusTrap = this.handleFocusTrap.bind(this)
+  this.boundKeyDown = this.handleKeyDown.bind(this)
+
+  // Modal elements
+  this.$buttonClose = this.$dialogBox.querySelector('.govuk-modal-dialogue__close')
+  this.$focussable = this.$dialogBox.querySelectorAll(this.focussable.toString())
+  this.$focusableLast = this.$focussable[this.$focussable.length - 1]
+  this.$focusElement = this.options.focusElement || this.$dialogBox
+
+  if (this.$dialogBox.hasAttribute('open')) {
+    this.open()
+  }
+
+  this.initEvents()
+
+  return this
+}
+
+// Initialize component events
+ModalDialogue.prototype.initEvents = function (options) {
+  if (this.options.triggerElement) {
+    this.options.triggerElement.addEventListener('click', this.open)
+  }
+
+  // Close dialogue on close button click
+  this.$buttonClose.addEventListener('click', this.close)
+}
+
+// Open modal
+ModalDialogue.prototype.handleOpen = function (event) {
+  if (event) {
+    event.preventDefault()
+  }
+
+  // Save last-focussed element
+  this.$lastActiveElement = document.activeElement
+
+  // Disable scrolling, show wrapper
+  this.$container.classList.add('govuk-!-scroll-disabled')
+  this.$module.classList.add('govuk-modal-dialogue--open')
+
+  // Close on escape key, trap focus
+  document.addEventListener('keydown', this.boundKeyDown, true)
+
+  // Optional 'onOpen' callback
+  if (typeof this.options.onOpen === 'function') {
+    this.options.onOpen.call(this)
+  }
+
+  // Skip open if already open
+  if (this.$dialogBox.hasAttribute('open')) {
+    return
+  }
+
+  // Show modal
+  this.hasNativeDialog
+    ? this.$dialogBox.show()
+    : this.$dialogBox.setAttribute('open', '')
+
+  // Handle focus
+  this.focus()
+}
+
+// Close modal
+ModalDialogue.prototype.handleClose = function (event) {
+  if (event) {
+    event.preventDefault()
+  }
+
+  // Skip close if already closed
+  if (!this.$dialogBox.hasAttribute('open')) {
+    return
+  }
+
+  // Hide modal
+  this.hasNativeDialog
+    ? this.$dialogBox.close()
+    : this.$dialogBox.removeAttribute('open')
+
+  // Hide wrapper, enable scrolling
+  this.$module.classList.remove('govuk-modal-dialogue--open')
+  this.$container.classList.remove('govuk-!-scroll-disabled')
+
+  // Restore focus to last active element
+  this.$lastActiveElement.focus()
+
+  // Optional 'onClose' callback
+  if (typeof this.options.onClose === 'function') {
```

Comment:
I wonder if we need to consider how we're documenting the JavaScript options for components, since we use data-attributes elsewhere instead of programmatic options.

See also: [Proposal programmatic and data attribute api interface for components](https://github.com/alphagov/govuk-frontend/pull/1238)

Reply:
@nickcolley Yeah good point.

Here's how we currently use this Modal API on our prototype:

```js
var $dialogue = document.querySelector('[data-module="govuk-modal-dialogue"]')
var $dialogueButtonResume = document.querySelector('.govuk-modal-dialogue__continue')

if ($dialogue && $dialogueButtonResume) {
  var modal = new window.ModalDialogue($dialogue)
    .init({
      focusElement: $dialogueButtonResume,
      onClose: function () {
        // Extend session timeout here
      }
    })
}
```